resources/qml/TimelineView.qml

```diff
@@ -497,6 +497,146 @@ Page {
 				}
 			}
 		}
+
+		Rectangle {
+			id: activeCallBar
+			visible: timelineManager.callState != WebRTCState.DISCONNECTED
+
+			Layout.fillWidth: true
+			implicitHeight: topLayout.height + 16
+			color: "#2ECC71"
+			z: 3
+
+			GridLayout {
+				anchors.left: parent.left
+				anchors.right: parent.right
+				anchors.margins: 8
+				anchors.verticalCenter: parent.verticalCenter
+
+				Avatar {
+					Layout.column: 1
+					Layout.row: 0
+					Layout.rowSpan: 2
+					Layout.alignment: Qt.AlignVCenter
+
+					width: avatarSize
+					height: avatarSize
+
+					url: chat.model ? chat.model.roomAvatarUrl.replace("mxc://", "image://MxcImage/") : ""
+					displayName: chat.model ? chat.model.roomName : qsTr("No room selected")
+				}
+
+				Label {
+					Layout.column: 2
+					Layout.row: 0
+					Layout.rowSpan: 2
+					Layout.alignment: Qt.AlignVCenter
+
+					font.pointSize: fontMetrics.font.pointSize * 1.1
+					text: chat.model ? "  " + chat.model.roomName + " " : ""
+				}
+
+				Image {
+					Layout.column: 3
+					Layout.row: 0
+					Layout.rowSpan: 2
+					Layout.alignment: Qt.AlignVCenter
+					Layout.preferredWidth: 23
+					Layout.preferredHeight: 23
+					source: "qrc:/icons/icons/ui/place-call.png"
+				}
+
+				Connections {
+					target: timelineManager
+					function onCallStateChanged(state) {
+						switch (state) {
+							case WebRTCState.INITIATING:
+								callStateLabel.text = "Initiating call..."
+								break;
+							case WebRTCState.INITIATED:
+								callStateLabel.text = "Call initiated..."
+								break;
+							case WebRTCState.OFFERSENT:
+								callStateLabel.text = "Calling..."
+								break;
+							case WebRTCState.CONNECTING:
+								callStateLabel.text = "Connecting..."
+								break;
+							case WebRTCState.CONNECTED:
+								callStateLabel.text = "00:00"
+								var d = new Date()
+								callTimer.startTime = Math.floor(d.getTime() / 1000)
+								break;
+							}
+					}
+				 }
+
+				Label {
+					id: callStateLabel
+					Layout.column: 4
+					Layout.row: 0
+					Layout.rowSpan: 2
+					Layout.alignment: Qt.AlignVCenter
+					font.pointSize: fontMetrics.font.pointSize * 1.1
+				}
+
+				Timer {
+					id: callTimer
+					property int startTime
+					interval: 1000
+					running: timelineManager.callState == WebRTCState.CONNECTED
+					repeat: true
+					onTriggered: {
+						var d = new Date()
+						let seconds = Math.floor(d.getTime() / 1000 - startTime)
+						let s = Math.floor(seconds % 60)
+						let m = Math.floor(seconds / 60) % 60
+						let h = Math.floor(seconds / 3600)
+						callStateLabel.text = (h ? (pad(h) + ":") : "") + pad(m) + ":" + pad(s)
```

Comment:
It's not that obvious, that the callStateLabel gets modified by the callTimer in some cases. This may be a bit more obvious using states, although I'm not too sure, if that solution is better (and if it works). Basically you add a string property for the elapsed time to the Timer and connect to that (which is not possible using a switch afaik).

```suggestion
				Label {
					id: callStateLabel
					Layout.column: 4
					Layout.row: 0
					Layout.rowSpan: 2
					Layout.alignment: Qt.AlignVCenter
					font.pointSize: fontMetrics.font.pointSize * 1.1
					
					states: [
					    State {
                                                name: "initiating"
                                                when: timelineManager.callState == WebRTCState.INITIATING
                                                PropertyChanges { target: callStateLabel; text: qsTr("Initiating call...") }
                                            },
					    State {
                                                name: "offersent"
                                                when: timelineManager.callState == WebRTCState.OFFERSENT
                                                PropertyChanges { target: callStateLabel; text: qsTr("Calling...") }
                                            },
					    State {
                                                name: "connecting"
                                                when: timelineManager.callState == WebRTCState.CONNECTING
                                                PropertyChanges { target: callStateLabel; text: qsTr("Connecting...") }
                                            },
					    State {
                                                name: "connected"
                                                when: timelineManager.callState == WebRTCState.CONNECTED
                                                PropertyChanges { target: callStateLabel; text: callTimer.elapsedTime }
                                                PropertyChanges { target: callTimer; startTime: (new Date()).getTime()/1000 }
                                            }
                                        ]
				}

				Timer {
					id: callTimer
					property int startTime
					property string elapsedTime: "00:00"
					interval: 1000
					running: timelineManager.callState == WebRTCState.CONNECTED
					repeat: true
					onTriggered: {
						var d = new Date()
						let seconds = Math.floor(d.getTime() / 1000 - startTime)
						let s = Math.floor(seconds % 60)
						let m = Math.floor(seconds / 60) % 60
						let h = Math.floor(seconds / 3600)
						elapsedTime = (h ? (pad(h) + ":") : "") + pad(m) + ":" + pad(s)
```

Reply:
I tried using States and it almost worked but I couldn't quite coerce it to do everything we need.

Reply:
I think this is fine then. It's localized to a single file now, so it should be easy enough to follow.